alf/algorithms/functional_particle_vi_algorithm.py

```diff
@@ -0,0 +1,518 @@
+            further processing.
+    Returns:
+        Tensor: the expaneded inputs or the original inputs.
+    """
+    outer_rank = get_outer_rank(inputs, spec)
+    if outer_rank == 1 and replicas > 1:
+        return inputs.unsqueeze(1).expand(-1, replicas, *inputs.shape[1:])
+    else:
+        return inputs
+
+
+@gin.configurable
+class FuncParVIAlgorithm(ParVIAlgorithm):
+    """Functional ParVI Algorithm 
+
+    Functional ParVI algorithm maintains a set of functional particles, 
+    where each particle is a neural network. All particles are updated
+    using particle-based VI approaches.
+
+    There are two ways of treating a neural network as a particle: 
+
+    * All the weights of the neural network as a particle.
+
+    * Outputs of the neural network for an input mini-batch as a particle.  
+
+    """
+
+    def __init__(self,
+                 input_tensor_spec=None,
+                 param_net: ParamNetwork = None,
+                 train_state_spec=(),
+                 conv_layer_params=None,
+                 fc_layer_params=None,
+                 activation=torch.relu_,
+                 last_layer_param=None,
+                 last_activation=None,
+                 num_particles=10,
+                 entropy_regularization=1.,
+                 loss_type="classification",
+                 voting="soft",
+                 par_vi="svgd",
+                 function_vi=False,
+                 function_bs=None,
+                 function_extra_bs_ratio=0.1,
+                 function_extra_bs_sampler='uniform',
+                 function_extra_bs_std=1.,
+                 optimizer=None,
+                 logging_network=False,
+                 logging_training=False,
+                 logging_evaluate=False,
+                 config: TrainerConfig = None,
+                 debug_summaries=False,
+                 name="FuncParVIAlgorithm"):
+        """
+        Args:
+            Args for the each parametric network
+            ====================================================================
+            input_tensor_spec (nested TensorSpec): the (nested) tensor spec of
+                the input. If nested, then ``preprocessing_combiner`` must not be
+                None.
+            param_net (ParamNetwork): input parametric network.
+            train_state_spec (nested TensorSpec): for the network state of
+                ``train_step()``.
+            conv_layer_params (tuple[tuple]): a tuple of tuples where each
+                tuple takes a format 
+                ``(filters, kernel_size, strides, padding, pooling_kernel)``,
+                where ``padding`` and ``pooling_kernel`` are optional.
+            fc_layer_params (tuple[tuple]): a tuple of tuples where each tuple
+                takes a format ``(FC layer sizes. use_bias)``, where 
+                ``use_bias`` is optional.
+            activation (nn.functional): activation used for all the layers but
```

Comment:
nn.functional is not a type.
activation (Callable)

Reply:
Updated.